app/src/test/java/org/simple/clinic/editpatient/PatientEditScreenControllerTest.kt

```diff
@@ -415,6 +415,13 @@ class PatientEditScreenControllerTest {
     verify(screen).showValidationErrors(setOf(expectedError))
   }
 
+  fun `params for date of birth should be validated`(): List<List<Any>> {
+    return listOf(
+        listOf("01/01/2000", InvalidPattern, INVALID_DATE_OF_BIRTH),
```

Comment:
Something I like to do whenever I need to add a method source for a parameterized test is to add a local `testCase` function so I can construct the list of parameters in a typed fashion and expose the intention using named parameters.

```kotlin
fun `params for date of birth should be validated`(): List<List<Any>> {
    fun testCase(
        dateOfBirth: String,
        validationResult: UserInputDateValidator.Result,
        expectedError: PatientEditValidationError
    ): List<Any> {
      return listOf(dateOfBirth, validationResult, expectedError)
    }

    return listOf(
        testCase(dateOfBirth = "01/01/2000", validationResult = InvalidPattern, expectedError = INVALID_DATE_OF_BIRTH),
        testCase(dateOfBirth = "01/01/2000", validationResult = DateIsInFuture, expectedError = DATE_OF_BIRTH_IN_FUTURE)
    )
}
```

IMO, this makes the test more maintainable. I'm not asking you to change this, just adding a suggestion about what I prefer to do.

Reply:
Would using data classes make this better, because it would be more type-safe?

Reply:
The problem will data classes will be that you'll need to convert them into a `List<Any>` again before you return from the method source.

Or the test function will need to accept the data class itself as the parameter. I personally feel that it's overkill, but we could try it out and see what it looks like.

Reply:
Used a data class.